aws/data_source_aws_vpc_ids.go

```diff
@@ -0,0 +1,63 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func dataSourceAwsVpcIDs() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceAwsVpcIDsRead,
+		Schema: map[string]*schema.Schema{
+			"filter": dataSourceFiltersSchema(),
+
+			"tags": tagsSchemaComputed(),
+
+			"ids": {
+				Type:     schema.TypeSet,
+				Computed: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+				Set:      schema.HashString,
+			},
+		},
+	}
+}
+
+func dataSourceAwsVpcIDsRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).ec2conn
+
+	req := &ec2.DescribeVpcsInput{}
+
+	req.Filters = buildEC2TagFilterList(
+		tagsFromMap(d.Get("tags").(map[string]interface{})),
+	)
+	if len(req.Filters) == 0 {
+		// Don't send an empty filters list; the EC2 API won't accept it.
+		req.Filters = nil
+	}
+
+	log.Printf("[DEBUG] DescribeVpcs %s\n", req)
+	resp, err := conn.DescribeVpcs(req)
+	if err != nil {
+		return err
+	}
+
+	if resp == nil || len(resp.Vpcs) == 0 {
+		return fmt.Errorf("no matching VPC found")
+	}
+
+	vpcs := make([]string, 0)
+
+	for _, vpc := range resp.Vpcs {
+		vpcs = append(vpcs, *vpc.VpcId)
+	}
+
+	d.SetId(time.Now().UTC().String())
+	d.Set("ids", vpcs)
```

Comment:
I would recommend adding error checking for setting a non-scalar attribute:

```go
if err := d.Set("ids", vpcs); err != nil {
  return fmt.Errorf("error setting ids: %s", err)
}
```

Reply:
Done, can you please explain about non-scalar attribute?

Reply:
Technically speaking, we should probably perform the error checking on all attributes during `d.Set()`, but its quite verbose code-wise and we have found nested and list/set attributes to be the most troublesome when it comes to ensuring that the type from the API matches the type defined in the schema (e.g. incorrectly trying to put `[]*string` into a `TypeList` with `TypeString` elements). Generally its worse when there is more attribute nesting happening when you need to be very concerned about lining up `[]map[string]interface{}` correctly.

When this type mismatch occurs, the behavior of Terraform core is to pass the configuration through into the state as-is from the configuration and not provide any feedback about this behavior unless the error check is in place. In the situation where we don't error check and provide incompatible type(s) into `d.Set()`, its not possible to detect drift from the API. It will also show as a plan difference after resource import since the state could not set the real value during the import read.

We have a flag in Terraform core to instead `panic()` when it runs into this situation (you can manually enable it today with `TF_SCHEMA_PANIC_ON_ERROR=1` environment variable), but I'm not sure we'll ever flip it on given the current state of many of the providers and generally poorer user experience that provides even though it would certainly make these bugs way more obvious.

Reply:
thanks